fuzzer/src/main.rs

```diff
@@ -1,78 +1,216 @@
+/// `round_trip_sort` will decode -> encode and compare the obtained encoded SORTED bytes with the original SORTED data.
+macro_rules! fuzz_decoder {
 	(
+		$fuzz_flow:ident;
 		$data:ident;
 		$first:ty,
 		$( $rest:ty, )*
 	) => {
-		fuzz_types! {
+		fuzz_decoder! {
 			@INTERNAL
+			$fuzz_flow;
 			$data;
 			1u8;
 			{ $first; 0u8 }
 			$( $rest, )*
 		}
 	};
 	(@INTERNAL
+		$fuzz_flow:ident;
 		$data:ident;
 		$counter:expr;
 		{ $( $parsed:ty; $index:expr ),* }
 		$current:ty,
 		$( $rest:ty, )*
 	) => {
-		fuzz_types! {
+		fuzz_decoder! {
 			@INTERNAL
+			$fuzz_flow;
 			$data;
 			$counter + 1u8;
 			{ $current; $counter $(, $parsed; $index )* }
 			$( $rest, )*
 		}
 	};
+	// round_trip flow arm.
+	(@INTERNAL
+		round_trip;
+		$data:ident;
+		$counter:expr;
+		{ $( $parsed:ty; $index:expr ),* }
+	) => {
+		let num = $counter;
+	$(
+		if $data[0] % num == $index {
+			let mut d = &$data[1..];
+			let raw1 = d.clone();
+			let maybe_obj = <$parsed>::decode(&mut d);
+
+			if let Ok(obj) = maybe_obj {
+				let mut d2: &[u8] = &obj.encode();
+				let raw2 = d2.clone();
+				let exp_obj = <$parsed>::decode(&mut d2);
+				match exp_obj {
+					Ok(obj2) => {
+						if obj == obj2 {
+							let raw1_trunc_to_obj_size = &raw1[..raw1.len()-d.len()];
+							if raw1_trunc_to_obj_size != raw2 {
+								println!("raw1 = {:?}", raw1);
+								println!("d (leftover/undecoded data) = {:?}", d);
+								println!("- Decoded data:");
+								println!("raw1_trunc = {:?}", raw1_trunc_to_obj_size);
+								println!("raw2 = {:?}", raw2);
+								println!("- Encoded objects:");
+								println!("obj1 = '{:?}'", obj);
+								println!("obj2 = '{:?}'", obj2);
+								println!("Type: {}", std::any::type_name::<$parsed>());
+								panic!("raw1 != raw2");
+							}
+						return
+						}
+					panic!("obj != obj2; obj={:?}, obj2={:?}", obj, obj2);
+					},
+					Err(e) => {
+						panic!("Shouldn’t happen: can't .decode() after .decode().encode(): {}", e);
+					}
+				}
+			}
+			return
+		}
+	)*
+	};
+	// only_decode flow arm.
 	(@INTERNAL
+		only_decode;
 		$data:ident;
 		$counter:expr;
 		{ $( $parsed:ty; $index:expr ),* }
 	) => {
 		let num = $counter;
 		$(
 			if $data[0] % num == $index {
-				// Check that decode doesn't panic.
+				// Check that decode doesn't panic
 				let _ = <$parsed>::decode(&mut &$data[1..]);
 				return
 			}
 		)*
+	};
+	// round_trip_sorted flow arm.
+	(@INTERNAL
+		round_trip_sorted;
+		$data:ident;
+		$counter:expr;
+		{ $( $parsed:ty; $index:expr ),* }
+	) => {
+		let num = $counter;
+	$(
+		if $data[0] % num == $index {
+			let mut d = &$data[1..];
+			let raw1 = &d.clone();
 
-		unreachable!()
+			let maybe_obj = <$parsed>::decode(&mut d);
+			if let Ok(obj) = maybe_obj {
+				let mut d2: &[u8] = &obj.encode();
+				let mut raw2 = Vec::from(d2);
+				// We are sorting here because we're in the "sorted" flow. Useful for container types
+				// which can have multiple valid encoded versions.
+				raw2.sort();
```

Comment:
You can not sort the raw data! That breaks the encoding!

Reply:
Yes it does indeed! This is what I tried to explain in the comments earlier: either we sort this (breaks encoding, but at least we can compare those bytes with the bytes we'll get after encoding the decoded.  OR, we can decide not to sort those, in which case we''l get false positives, so we should probably just not add those types to the decode-encode flow (and just try and decode them, period).

The first solution is a superset of the second solution (it decodes + encodes) so I would argue that we should use it whenever we can. Of course it is not perfect because we're not comparing the bytes directly but the sorted bytes, but hey it's better than nothing right? :p